internal/k8s/cache/gateway.go

```diff
@@ -0,0 +1,75 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gatewayapi_v1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
+)
+
+type gatewayReconciler struct {
+	client       client.Client
+	eventHandler cache.ResourceEventHandler
+	logrus.FieldLogger
+}
+
+// NewGatewayController creates the gateway controller from mgr. The controller will be pre-configured
+// to watch for Gateway objects across all namespaces.
+func NewGatewayController(mgr manager.Manager, eventHandler cache.ResourceEventHandler, log logrus.FieldLogger) (controller.Controller, error) {
+	r := &gatewayReconciler{
+		client:       mgr.GetClient(),
+		eventHandler: eventHandler,
+		FieldLogger:  log,
+	}
+	c, err := controller.New("gateway-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return nil, err
+	}
+	if err := c.Watch(&source.Kind{Type: &gatewayapi_v1alpha1.Gateway{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (r *gatewayReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	// Fetch the Gateway from the cache.
+	gateway := &gatewayapi_v1alpha1.Gateway{}
+	err := r.client.Get(ctx, request.NamespacedName, gateway)
+	if errors.IsNotFound(err) {
+		r.Error(nil, "Could not find Gateway %q in Namespace %q", request.Name, request.Namespace)
+		return reconcile.Result{}, nil
+	}
```

Comment:
should this case result in an `OnDelete` event?

Reply:
Nope it shouldn't, this handles that the event fired but can't be found in the internal cache. I think once we get the finalizers bit it will mean more since the object can't be removed until those are dropped.

Reply:
OK, yeah I agree on the finalizers. My assumption was that if an object (without finalizers) was deleted, then `Reconcile` would get triggered but since the object was deleted, it would potentially not be in the cache (i.e. this case). In that scenario, we'd want to make sure we removed the item from the Contour cache. But if everything's got finalizers, then the `deletionTimestamp` case should handle it.

Reply:
I can add it to delete just to make sure, but would want to extend the delete method in the cache.go to check if the object exists first in the event that it got removed, but then this got called again for some reason.

Reply:
Just looked in the kubebuilder book (ref. https://book.kubebuilder.io/reference/using-finalizers.html):

> The key point to note is that a finalizer causes “delete” on the object to become an “update” to set deletion timestamp. Presence of deletion timestamp on the object indicates that it is being deleted. **Otherwise, without finalizers, a delete shows up as a reconcile where the object is missing from the cache.**

Reply:
...or maybe should remove the timestamp checks since we don't have finalizers?